torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -9,19 +11,39 @@ namespace fuser {
 namespace cuda {
 
 at::optional<CudaKernel*> CudaKernelCache::getKernelPtr(
-    const at::ArrayRef<c10::IValue> inputs) {
-  for (auto& iter : kernels_) {
-    if (iter.first->matchKernelSize(inputs)) {
-      return &(iter.second);
+    const at::ArrayRef<c10::IValue> inputs,
+    const std::vector<int64_t>& broadcasted_shape) {
+  for (auto& cuda_kernel : kernels_) {
+    // bound input sizes
+    Fusion* fusion = cuda_kernel.getFusionPtr();
+    FusionGuard fg(fusion);
+    EvaluationContext eval_context(fusion);
+    for (int i = 0; i < (int)inputs.size(); i++) {
+      if (inputs[i].isTensor()) {
+        ExtractSizeStride ess(inputs[i].toTensor(), broadcasted_shape);
+        int nDims = ess.sizes.size();
+        TensorView* tv = fusion->inputs()[i]->as<TensorView>();
+        for (int j = 0; j < nDims; j++) {
+          eval_context.bind(tv->getRootDomain()[j]->extent(), ess.sizes[j]);
+        }
+      }
+    }
+
+    const auto val = ExpressionEvaluator::evaluate(
+        fusion->getLaunchConfig(LaunchConfigType::Compatible), &eval_context);
+    TORCH_INTERNAL_ASSERT(
+        val.has_value(), "scheduler didn't bind launch configs properly");
+    if (val.value()) {
+      return &(cuda_kernel);
     }
   }
   return at::nullopt;
 }
 
 CudaKernel* CudaKernelCache::allocateKernelInCache(
-    std::unique_ptr<KernelArgsReq>&& args_req) {
-  kernels_.emplace_back(std::move(args_req), CudaKernel());
-  return &(kernels_.back().second);
+    const at::ArrayRef<c10::IValue> inputs) {
+  kernels_.emplace_back();
```

Comment:
emplace back what?

Reply:
emplace_back with empty argument, so pushing back an element with default constructor. I can be more explicit and use `kernels_.emplace_back(CudaKernel())`.

I've been skipping the argument quite often just to get shorter code, is it a bad style?

Reply:
I wanted to make sure it's intentional.

Personally I'd prefer the more explicit push_back(CudaKernel()), but that's a matter of taste. Note that emplace_back(CudaKernel()) doesn't really buy you anything compared to push_back() - both will use the copy constructor.

Reply:
> emplace_back(CudaKernel()) doesn't really buy you anything compared to push_back() - both will use the copy constructor.
Agree that they behave the same in this case and they both use move constructor, but that's not saving anything.

emplace_back() is different though, as you construct in-place and there's not copy.